src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs

```diff
@@ -0,0 +1,182 @@
+                    EventIds.ResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static IDisposable? BeginRequestPipelineScope(this ILogger logger, HttpRequestMessage request)
+        {
+            return _beginRequestPipelineScope(logger, request.Method, GetRedactedUriString(request.RequestUri));
+        }
+
+        public static void LogRequestPipelineStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineRequestHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Request, request.Headers, request.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static void LogRequestPipelineEnd(this ILogger logger, HttpResponseMessage response, TimeSpan duration, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineEnd(logger, duration.TotalMilliseconds, (int)response.StatusCode, null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        internal static string? GetRedactedUriString(Uri? uri)
+        {
+            if (uri is null)
+            {
+                return null;
+            }
+
+            string uriString = uri.IsAbsoluteUri ? uri.AbsoluteUri : uri.ToString();
+
+            if (s_logQueryString)
+            {
+                return uriString;
+            }
+
+            int queryOffset = uriString.IndexOf('?');
+
+            if (queryOffset < 0 || queryOffset == uriString.Length - 1)
+            {
+                // No query string or empty query string.
+                return uriString;
+            }
+
+#if NET
+            return string.Create(queryOffset + 2, (uriString, queryOffset), (result, s) =>
+            {
+                ReadOnlySpan<char> withoutQuery = s.uriString.AsSpan(0, s.queryOffset + 1);
+                withoutQuery.CopyTo(result);
+                Debug.Assert(result[^2] == '?');
+                result[^1] = '*';
+            });
```

Comment:
```suggestion
            return $"{uriString.AsSpan(0, queryOffset + 1)}*";
```

We could probably come up with some incantation of `$"{uri.Scheme}://{uri.Foo}{someSpan}*"` to save 1 allocation if we care enough.